### ISSUE TITLE
Crash with Proxy Protocol v2 LOCAL packet

### ISSUE THREAD

**Issue by justanotherboy** [rabbitmq/rabbitmq-server#8654]
### Describe the bug

child is terminated when proxy_protocol is enabled and a [Proxy Protocol v2 LOCAL packet](https://github.com/haproxy/haproxy/blob/master/doc/proxy-protocol.txt#L346) is received with length greater than zero.

The error message is:

```
2023-06-21 17:33:07.977016+00:00 [error] <0.1817.0>   crasher:
2023-06-21 17:33:07.977016+00:00 [error] <0.1817.0>     initial call: rabbit_reader:init/3
2023-06-21 17:33:07.977016+00:00 [error] <0.1817.0>     pid: <0.1817.0>
2023-06-21 17:33:07.977016+00:00 [error] <0.1817.0>     registered_name: []
2023-06-21 17:33:07.977016+00:00 [error] <0.1817.0>     exception error: no match of right hand side value #{command => local,
2023-06-21 17:33:07.977016+00:00 [error] <0.1817.0>                                                          version => 2}
2023-06-21 17:33:07.977016+00:00 [error] <0.1817.0>       in function  rabbit_net:socket_ends/2 (rabbit_net.erl, line 273)
2023-06-21 17:33:07.977016+00:00 [error] <0.1817.0>       in call from rabbit_net:connection_string/2 (rabbit_net.erl, line 250)
2023-06-21 17:33:07.977016+00:00 [error] <0.1817.0>       in call from rabbit_reader:start_connection/5 (rabbit_reader.erl, line 300)
2023-06-21 17:33:07.977016+00:00 [error] <0.1817.0>     ancestors: [<0.1815.0>,<0.688.0>,<0.687.0>,<0.686.0>,<0.684.0>,
2023-06-21 17:33:07.977016+00:00 [error] <0.1817.0>                   <0.683.0>,rabbit_sup,<0.235.0>]
2023-06-21 17:33:07.977016+00:00 [error] <0.1817.0>     message_queue_len: 0
2023-06-21 17:33:07.977016+00:00 [error] <0.1817.0>     messages: []
2023-06-21 17:33:07.977016+00:00 [error] <0.1817.0>     links: [<0.1815.0>,#Port<0.426>]
2023-06-21 17:33:07.977016+00:00 [error] <0.1817.0>     dictionary: []
2023-06-21 17:33:07.977016+00:00 [error] <0.1817.0>     trap_exit: true
2023-06-21 17:33:07.977016+00:00 [error] <0.1817.0>     status: running
2023-06-21 17:33:07.977016+00:00 [error] <0.1817.0>     heap_size: 376
2023-06-21 17:33:07.977016+00:00 [error] <0.1817.0>     stack_size: 28
2023-06-21 17:33:07.977016+00:00 [error] <0.1817.0>     reductions: 485
2023-06-21 17:33:07.977016+00:00 [error] <0.1817.0>   neighbours:
2023-06-21 17:33:07.977016+00:00 [error] <0.1817.0>
2023-06-21 17:33:07.977445+00:00 [error] <0.1815.0>     supervisor: {<0.1815.0>,rabbit_connection_sup}
2023-06-21 17:33:07.977445+00:00 [error] <0.1815.0>     errorContext: child_terminated
2023-06-21 17:33:07.977445+00:00 [error] <0.1815.0>     reason: {{badmatch,#{command => local,version => 2}},
2023-06-21 17:33:07.977445+00:00 [error] <0.1815.0>              [{rabbit_net,socket_ends,2,[{file,"rabbit_net.erl"},{line,273}]},
2023-06-21 17:33:07.977445+00:00 [error] <0.1815.0>               {rabbit_net,connection_string,2,
2023-06-21 17:33:07.977445+00:00 [error] <0.1815.0>                           [{file,"rabbit_net.erl"},{line,250}]},
2023-06-21 17:33:07.977445+00:00 [error] <0.1815.0>               {rabbit_reader,start_connection,5,
2023-06-21 17:33:07.977445+00:00 [error] <0.1815.0>                              [{file,"rabbit_reader.erl"},{line,300}]},
2023-06-21 17:33:07.977445+00:00 [error] <0.1815.0>               {proc_lib,init_p_do_apply,3,
2023-06-21 17:33:07.977445+00:00 [error] <0.1815.0>                         [{file,"proc_lib.erl"},{line,241}]}]}
2023-06-21 17:33:07.977445+00:00 [error] <0.1815.0>     offender: [{pid,<0.1817.0>},
2023-06-21 17:33:07.977445+00:00 [error] <0.1815.0>                {id,reader},
2023-06-21 17:33:07.977445+00:00 [error] <0.1815.0>                {mfargs,{rabbit_reader,start_link,
2023-06-21 17:33:07.977445+00:00 [error] <0.1815.0>                                       [<0.1816.0>,
2023-06-21 17:33:07.977445+00:00 [error] <0.1815.0>                                        {acceptor,{0,0,0,0,0,0,0,0},5672}]}},
2023-06-21 17:33:07.977445+00:00 [error] <0.1815.0>                {restart_type,transient},
2023-06-21 17:33:07.977445+00:00 [error] <0.1815.0>                {significant,true},
2023-06-21 17:33:07.977445+00:00 [error] <0.1815.0>                {shutdown,300000},
2023-06-21 17:33:07.977445+00:00 [error] <0.1815.0>                {child_type,worker}]
2023-06-21 17:33:07.977445+00:00 [error] <0.1815.0>
2023-06-21 17:33:07.978002+00:00 [error] <0.1815.0>     supervisor: {<0.1815.0>,rabbit_connection_sup}
2023-06-21 17:33:07.978002+00:00 [error] <0.1815.0>     errorContext: shutdown
2023-06-21 17:33:07.978002+00:00 [error] <0.1815.0>     reason: reached_max_restart_intensity
2023-06-21 17:33:07.978002+00:00 [error] <0.1815.0>     offender: [{pid,<0.1817.0>},
2023-06-21 17:33:07.978002+00:00 [error] <0.1815.0>                {id,reader},
2023-06-21 17:33:07.978002+00:00 [error] <0.1815.0>                {mfargs,{rabbit_reader,start_link,
2023-06-21 17:33:07.978002+00:00 [error] <0.1815.0>                                       [<0.1816.0>,
2023-06-21 17:33:07.978002+00:00 [error] <0.1815.0>                                        {acceptor,{0,0,0,0,0,0,0,0},5672}]}},
2023-06-21 17:33:07.978002+00:00 [error] <0.1815.0>                {restart_type,transient},
2023-06-21 17:33:07.978002+00:00 [error] <0.1815.0>                {significant,true},
2023-06-21 17:33:07.978002+00:00 [error] <0.1815.0>                {shutdown,300000},
2023-06-21 17:33:07.978002+00:00 [error] <0.1815.0>                {child_type,worker}]
```

### Reproduction steps

1. Set `proxy_protocol = true` in the configuration file
2. Start the cluster
3. Send a Proxy Protocol v2 LOCAL packet with [PP2_TYPE_CRC32C](https://github.com/haproxy/haproxy/blob/master/doc/proxy-protocol.txt#L569) (Length set to 0x0007 and then the CRC)

### Expected behavior

- The packet must not crash the application
- The rest of the packet can be ignored per the [proxy protocol specification](https://github.com/haproxy/haproxy/blob/master/doc/proxy-protocol.txt#L514)

### Additional context

- RabbitMQ version 3.12.0
- Erlang version 26.0
- rabbitmq.conf:
```
## Username and password
##
default_user = XXX
default_pass = $(RABBITMQ_PASSWORD)

# queue master locator
queue_master_locator = min-masters
# enable loopback user
loopback_users.dev = false

# The requests to the service go through AWS ELB, we need to enable proxy protocol
proxy_protocol = true

## Prometheus metrics
##
prometheus.tcp.port = 9419
```

- The data in the proxy protocol that breaks the application:

```
0x0d 0x0a 0x0d 0x0a 0x00
0x0d 0x0a 0x51 0x55 0x49
0x54 0x0a 0x20 0x00 0x00
0x07 0x03 0x00 0x04 0xa9
0xb8 0x7e 0x8f
```

I can provide the packet capture if necessary

**Comment by michaelklishin** [comment]
Thank you for the details. A traffic capture would be great to have, because it's less than obvious to me how (besides `nc` or a tiny custom program) exactly to perform step 3.

**Comment by justanotherboy** [comment]
Here is the packet capture (it is zipped so github doesn't complain). I filtered the capture to a single TCP session for easier visualization but the same behavior happens over and over again.

You can see in the packet number 4 the PROXY protocol header that is crashing the application.

Let me know if I can provide something else for the debugging.

[report.pcap.zip](https://github.com/rabbitmq/rabbitmq-server/files/11837349/report.pcap.zip)

**Comment by lhoguin** [comment]
Rabbit uses Ranch, which parses the packet just fine, but Rabbit does not currently support LOCAL packets. If you can I would suggest not using them for now. I will take a look tomorrow at adding support for them.

**Comment by justanotherboy** [comment]
Thank you (both of you) for your fast response. These LOCAL packets are used for the health checks in AWS ELB (which can't be disabled). I will redirect these probes to another open port in the same host to avoid crashing the application meanwhile.

**Comment by michaelklishin** [comment]
Addressed in #8664.

**Comment by lhoguin** [comment]
Thank you for the report!
